[SPARK-12656] [SQL] Implement Intersect with Left-semi Join

gatorsmile opened the pull request:
Our current Intersect physical operator simply delegates to RDD.intersect. We should remove the Intersect physical operator and simply transform a logical intersect into a semi-join with distinct. This way, we can take advantage of all the benefits of join implementations (e.g. managed memory, code generation, broadcast joins).

After a search, I found one of the mainstream RDBMS did the same. In their query explain, Intersect is replaced by Left-semi Join. Left-semi Join could help outer-join elimination in Optimizer, as shown in the PR: https://github.com/apache/spark/pull/10566

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
SPARK-11633

gatorsmile committed:
Merge remote-tracking branch 'upstream/master' into joinMakeCopy

gatorspark committed:
Merge remote-tracking branch 'upstream/master'

gatorspark committed:
Merge branch 'master' of https://github.com/gatorsmile/spark

Commit:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
converge

gatorsmile committed:
converge

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
replace Intersect with Left-semi Join

gatorsmile commented:
@rxin Please review the implementation. Thank you!

rxin commented:
Which mainstream RDBMS is that?

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -966,6 +952,27 @@ object ReplaceDistinctWithAggregate extends Rule[LogicalPlan] {
+ *   ==>  SELECT a1, a2 FROM Tab1, Tab2 ON a1<=>b1 AND a2<=>b2
+ * }}}
+ */
+object ReplaceIntersectWithLeftSemi extends Rule[LogicalPlan] {
```
LeftSemiJoin

gatorsmile replied:
yeah. Forgot to specify the join type

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -966,6 +952,27 @@ object ReplaceDistinctWithAggregate extends Rule[LogicalPlan] {
+ */
+object ReplaceIntersectWithLeftSemi extends Rule[LogicalPlan] {
+  private def buildCond (left: LogicalPlan, right: LogicalPlan): Seq[Expression] = {
+    require(left.output.length == right.output.length)
```
just inline this into the rule, since it is used only once?

rxin replied:
can you look into the analyzer to make sure we have the proper check there for intersect/except?

gatorsmile replied:
Yeah, will do it

gatorsmile replied:
Yeah, it has a checking logic in analyzer:
https://github.com/apache/spark/blob/d084a2de3271fd8b0d29ee67e1e214e8b9d96d6d/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala#L186-L190

gatorsmile commented:
MS SQL Server did that

rxin commented on sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:
```diff
@@ -322,13 +323,32 @@ class DataFrameSuite extends QueryTest with SharedSQLContext {
   test("intersect") {
+    val intersectDF = lowerCaseData.intersect(lowerCaseData)
+
+    // Before Optimizer, the operator is Intersect
```
this should go into one of the optimizer unit test suite, not here.

gatorsmile replied:
ok, will add a new test suite for it.

rxin commented on sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:
```diff
@@ -322,13 +323,32 @@ class DataFrameSuite extends QueryTest with SharedSQLContext {
       Row(4, "d") :: Nil)
     checkAnswer(lowerCaseData.intersect(upperCaseData), Nil)
+
+    checkAnswer(
```
leave a line of comment on what this is testing (e.g. null equality).

gatorsmile replied:
ok, will do it.

rxin commented:
LGTM. 

cc @cloud-fan to take a look too.

SparkQA commented:
**[Test build #48900 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48900/consoleFull)** for PR 10630 at commit [`0bd1771`](https://github.com/apache/spark/commit/0bd177135bc1c8210fa0ea5a8c7a4c5144d6227e).

SparkQA commented:
**[Test build #48900 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48900/consoleFull)** for PR 10630 at commit [`0bd1771`](https://github.com/apache/spark/commit/0bd177135bc1c8210fa0ea5a8c7a4c5144d6227e).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/48900/
Test FAILed.

gatorsmile committed:
Merge remote-tracking branch 'upstream/master' into IntersectBySemiJoin

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -966,6 +952,27 @@ object ReplaceDistinctWithAggregate extends Rule[LogicalPlan] {
+    }
+  }
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
```
use transformUp?

cc @yhuai

rxin replied:
actually nvm.

gatorsmile committed:
address comments.

gatorsmile committed:
Merge remote-tracking branch 'upstream/master' into IntersectBySemiJoin

gatorsmile committed:
clean code.

gatorsmile committed:
clean code.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -966,6 +952,22 @@ object ReplaceDistinctWithAggregate extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case Intersect(left, right) =>
+      val joinCond = left.output.zip(right.output).map { case (l, r) =>
+        EqualNullSafe(l, r) }
```
nit: can we put it in one line?

cloud-fan commented:
LGTM, pending tests.

It will be good to have some performance numbers for this change :)

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -966,6 +952,22 @@ object ReplaceDistinctWithAggregate extends Rule[LogicalPlan] {
+object ReplaceIntersectWithSemiJoin extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case Intersect(left, right) =>
+      val joinCond = left.output.zip(right.output).map { case (l, r) =>
```
one nit - might as well add assert(left.output.size == right.output.size)

gatorsmile replied:
Added. Thanks!

gatorsmile committed:
address comments.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/48907/
Test FAILed.

gatorsmile commented:
retest this please.

rxin commented:
LGTM.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/48909/
Test FAILed.

SparkQA commented:
**[Test build #2342 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2342/consoleFull)** for PR 10630 at commit [`6742984`](https://github.com/apache/spark/commit/6742984f5ec0d480435988712e161838c3e8d6d5).

SparkQA commented:
**[Test build #48912 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48912/consoleFull)** for PR 10630 at commit [`6742984`](https://github.com/apache/spark/commit/6742984f5ec0d480435988712e161838c3e8d6d5).

gatorsmile commented:
@cloud-fan Sorry, I am still requesting a resource to do the performance testing. Unfortunately, I am unable to get it very recently. I am also highly interested in the performance number when running a large-scale workload in a distributed environment. Will keep you posted if it is done. Thanks!

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/48911/
Test FAILed.

nongli commented on sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:
```diff
@@ -329,6 +329,14 @@ class DataFrameSuite extends QueryTest with SharedSQLContext {
       Row(4, "d") :: Nil)
     checkAnswer(lowerCaseData.intersect(upperCaseData), Nil)
+
+    // check null equality
```
Do we have a test case where there are duplicates on the left side?

nongli replied:
A good case might be two tables that each contain two int rows [1], [1] and intersect them.

gatorsmile replied:
Thank you! Let me add the test case. It passes but we should still add this case!

gatorsmile committed:
added one more case for duplicate values

SparkQA commented:
**[Test build #48916 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48916/consoleFull)** for PR 10630 at commit [`e4c34f0`](https://github.com/apache/spark/commit/e4c34f01acbb626a8084c99a5f0ad26fbfb175e2).

SparkQA commented:
**[Test build #48916 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48916/consoleFull)** for PR 10630 at commit [`e4c34f0`](https://github.com/apache/spark/commit/e4c34f01acbb626a8084c99a5f0ad26fbfb175e2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/48916/
Test PASSed.

SparkQA commented:
**[Test build #48912 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48912/consoleFull)** for PR 10630 at commit [`6742984`](https://github.com/apache/spark/commit/6742984f5ec0d480435988712e161838c3e8d6d5).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/48912/
Test FAILed.

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile commented:
retest this please

SparkQA commented:
**[Test build #48947 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48947/consoleFull)** for PR 10630 at commit [`e4c34f0`](https://github.com/apache/spark/commit/e4c34f01acbb626a8084c99a5f0ad26fbfb175e2).

yhuai commented on sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:
```diff
@@ -329,6 +329,22 @@ class DataFrameSuite extends QueryTest with SharedSQLContext {
+      Row(null) ::
+      Row(null) ::
+      Row(null) ::
+      Row(null) :: Nil)
```
Is the expected answer right?

 Also, can you also add the test mentioned by @nongli (handling null is kind of special. So let's also have a regular case).

gatorsmile replied:
Let me add it to the bucket. Thanks!

yhuai commented:
How about we hold on it a little bit? I am not sure the plan is right.

yhuai commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -966,6 +952,22 @@ object ReplaceDistinctWithAggregate extends Rule[LogicalPlan] {
+    case Intersect(left, right) =>
+      assert(left.output.size == right.output.size)
+      val joinCond = left.output.zip(right.output).map { case (l, r) => EqualNullSafe(l, r) }
+      Join(left, right, LeftSemi, joinCond.reduceLeftOption(And))
```
Looks like intersect also implies de-dup unless `INTERSECT ALL` is used (see http://www.postgresql.org/docs/9.4/static/queries-union.html and https://msdn.microsoft.com/en-us/library/ms188055.aspx).

gatorsmile replied:
You are right! It should remove duplicates. I just tried it in DB2.

yhuai replied:
Does our current intersect do de-dup?






On Thu, Jan 7, 2016 at 8:50 AM -0800, "Xiao Li" <notifications@github.com> wrote:












In sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
> @@ -966,6 +952,22 @@ object ReplaceDistinctWithAggregate extends Rule[LogicalPlan] {
>  }
>  
>  /**
> + * Replaces logical [[Intersect]] operator with a left-semi [[Join]] operator.
> + * {{{
> + *   SELECT a1, a2 FROM Tab1 INTERSECT SELECT b1, b2 FROM Tab2
> + *   ==>  SELECT a1, a2 FROM Tab1 LEFT SEMI JOIN Tab2 ON a1<=>b1 AND a2<=>b2
> + * }}}
> + */
> +object ReplaceIntersectWithSemiJoin extends Rule[LogicalPlan] {
> +  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
> +    case Intersect(left, right) =>
> +      assert(left.output.size == right.output.size)
> +      val joinCond = left.output.zip(right.output).map { case (l, r) => EqualNullSafe(l, r) }
> +      Join(left, right, LeftSemi, joinCond.reduceLeftOption(And))



You are right! It should remove duplicates. I just tried it in DB2. 



—
Reply to this email directly or view it on GitHub.

gatorsmile replied:
Just ran it. The existing `Intersect` removes the duplicates.

gatorsmile replied:
I also checked the underlying RDD API: `Intersect`. It removes the duplicates. If we added `Distinct` above `Intersect`, we need to measure the performance difference, I think.

rxin replied:
You can just insert a distinct into this.

gatorsmile replied:
Sure. Will do

gatorsmile commented:
Should we provide users one more Dataframe/SQL API `INTERSECT ALL`, which can be converted to `Left-semi Join`? Thank you! @yhuai @rxin @nongli @cloud-fan @marmbrus

SparkQA commented:
**[Test build #48947 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48947/consoleFull)** for PR 10630 at commit [`e4c34f0`](https://github.com/apache/spark/commit/e4c34f01acbb626a8084c99a5f0ad26fbfb175e2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

gatorsmile commented:
Also did a research in the SQL standard. Below is the SQL2003 Syntax
```
<SELECT statement1>
INTERSECT [ALL | DISTINCT]
[CORRESPONDING [BY (column1, column2, ...)]] <SELECT statement2>
INTERSECT [ALL | DISTINCT]
[CORRESPONDING [BY (column1, column2, ...)]] 
...
```

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/48947/
Test PASSed.

rxin commented:
It's reasonable to support intersect all. We should do that as a separate ticket though.

gatorsmile commented:
After the merge, will submit another PR to support it. Thanks!

gatorsmile committed:
added an exception in conversion from logical to physical operators.

gatorsmile committed:
Add DISTINCT and test cases.

gatorsmile committed:
test case updates.

SparkQA commented:
**[Test build #49007 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49007/consoleFull)** for PR 10630 at commit [`27192be`](https://github.com/apache/spark/commit/27192be27708265aba6707d1100eefca537fb2b8).

gatorsmile commented:
In this update, the code changes include
  - Added `DISTINCT` on top of `Semi Join` for `Intersect`
  - Added a test case for detecting if the duplicate values are removed after `Intersect`
  - Threw an exception if the `Intersect` is not converted before conversion from logical plan to physical plan
  - Merge two rules `ReplaceIntersectWithSemiJoin` and `ReplaceDistinctWithAggregate` in the same batch `"Replace Operators"` for ensuring `ReplaceIntersectWithSemiJoin` should be called before `ReplaceDistinctWithAggregate`

SparkQA commented:
**[Test build #49007 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49007/consoleFull)** for PR 10630 at commit [`27192be`](https://github.com/apache/spark/commit/27192be27708265aba6707d1100eefca537fb2b8).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49007/
Test FAILed.

gatorsmile commented:
retest this please

SparkQA commented:
**[Test build #49008 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49008/consoleFull)** for PR 10630 at commit [`27192be`](https://github.com/apache/spark/commit/27192be27708265aba6707d1100eefca537fb2b8).

SparkQA commented:
**[Test build #49008 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49008/consoleFull)** for PR 10630 at commit [`27192be`](https://github.com/apache/spark/commit/27192be27708265aba6707d1100eefca537fb2b8).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49008/
Test FAILed.

gatorsmile commented:
Both failure are not related to this PR. They are caused by the test case `randomSplit on reordered partitions` in https://github.com/apache/spark/pull/10626. Thanks!

gatorsmile committed:
Merge remote-tracking branch 'upstream/master' into IntersectBySemiJoin

SparkQA commented:
**[Test build #49012 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49012/consoleFull)** for PR 10630 at commit [`a932cdb`](https://github.com/apache/spark/commit/a932cdb05c889ded8de72024b2eef19723bdb0f4).

SparkQA commented:
**[Test build #49012 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49012/consoleFull)** for PR 10630 at commit [`a932cdb`](https://github.com/apache/spark/commit/a932cdb05c889ded8de72024b2eef19723bdb0f4).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class JavaSampleActorReceiver<T> extends JavaActorReceiver `
  * `public class JavaActorWordCount `
  * `abstract class ActorReceiver extends Actor `
  * `abstract class JavaActorReceiver extends UntypedActor `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49012/
Test FAILed.

gatorsmile commented:
Found the root cause of this failure. We also need to change the analyzer to handle duplicate expression ids.

gatorsmile committed:
resolve the ambiguous attributes

gatorsmile commented:
The latest update is for resolving the ambiguous attributes, which could trigger invalid predicate pushdown by the Optimizer rule `PushPredicateThroughJoin`. The solution is the same as what we are doing for self join in analyzer. 

Thank you!

SparkQA commented:
**[Test build #49020 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49020/consoleFull)** for PR 10630 at commit [`04a26bd`](https://github.com/apache/spark/commit/04a26bd3beaab9d8ea3a4b300eb9f413b5eaa704).

SparkQA commented:
**[Test build #49020 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49020/consoleFull)** for PR 10630 at commit [`04a26bd`](https://github.com/apache/spark/commit/04a26bd3beaab9d8ea3a4b300eb9f413b5eaa704).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class Intersect(left: LogicalPlan, right: LogicalPlan) extends SetOperation(left, right) `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49020/
Test PASSed.

marmbrus commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -388,57 +441,14 @@ class Analyzer(
+      // Special handling for cases when self-intersect introduces duplicate expression IDs.
+      // In Optimizer, Intersect will be converted to semi join.
+      case i @ Intersect(left, right) if !i.duplicateResolved =>
+        i.copy(right = buildRightChild4Deduplication(left, right))
```
Do we need to do this for all non-leaf/unary nodes?  I think the only reason this wasn't a problem in the past was there were no optimizer rules for those nodes.

gatorsmile replied:
I see, will do a change to make it more general. Thanks!

marmbrus commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -344,6 +344,59 @@ class Analyzer(
       }
     }
 
+    def buildRightChild4Deduplication (left: LogicalPlan, right: LogicalPlan): LogicalPlan = {
```
extra space.  Also I'm not a huge fan of `4`.  How about `dedupRight` with some scala doc.

gatorsmile replied:
Sure. : )

marmbrus commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala:
```diff
@@ -116,7 +116,18 @@ case class Union(left: LogicalPlan, right: LogicalPlan) extends SetOperation(lef
+  def duplicateResolved: Boolean = left.outputSet.intersect(right.outputSet).isEmpty
+
+  // Intersect is only resolved if they don't introduce ambiguous expression ids,
+  // since it will be converted to semi Join by optimizer.
```
This check might belong more generally in `childrenResolved` as I mention above.

gatorsmile replied:
Will update the comments in the code.

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge branch 'IntersectBySemiJoin' into IntersectBySemiJoinMerged

# Conflicts:
#	sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

gatorsmile committed:
resolve comments.

SparkQA commented:
**[Test build #49038 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49038/consoleFull)** for PR 10630 at commit [`f820c61`](https://github.com/apache/spark/commit/f820c616fe217494ccaed0bf74a0a7410ce503cf).

gatorsmile commented:
In this update, the code changes include
  - Fixed a bug in `Range` by inheriting the trait `MultiInstanceRelation`.
  - Added a de-duplication resolution for all the binary nodes: `Except`, `Union` and `Co-Group`, besides `Intersect` and `Join`.
  - Added a new function `duplicateResolved` for all the binary nodes.
  - Improved the analysis exception message when failure to resolve conflicting references
  - Resolved all the other comments. 

The analysis procedure is kind of tricky. I am unable to directly include `duplicateResolved` into `childrenResolved`. `resolve` is lazy evaluated. The resolution procedure need to follow the order: children at first, then itself, and then deduplicate the attributes' expression IDs in tis children.

gatorsmile committed:
style change.

SparkQA commented:
**[Test build #49039 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49039/consoleFull)** for PR 10630 at commit [`4372170`](https://github.com/apache/spark/commit/4372170f600eb25996c3aa4f09d569312c263686).

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

SparkQA commented:
**[Test build #49039 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49039/consoleFull)** for PR 10630 at commit [`4372170`](https://github.com/apache/spark/commit/4372170f600eb25996c3aa4f09d569312c263686).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49039/
Test PASSed.

SparkQA commented:
**[Test build #49038 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49038/consoleFull)** for PR 10630 at commit [`f820c61`](https://github.com/apache/spark/commit/f820c616fe217494ccaed0bf74a0a7410ce503cf).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class Intersect(left: LogicalPlan, right: LogicalPlan) extends SetOperation(left, right)`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49038/
Test FAILed.

gatorsmile commented:
retest this please

SparkQA commented:
**[Test build #49044 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49044/consoleFull)** for PR 10630 at commit [`4372170`](https://github.com/apache/spark/commit/4372170f600eb25996c3aa4f09d569312c263686).

SparkQA commented:
**[Test build #49044 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49044/consoleFull)** for PR 10630 at commit [`4372170`](https://github.com/apache/spark/commit/4372170f600eb25996c3aa4f09d569312c263686).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49044/
Test PASSed.

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

rxin commented:
@gatorsmile there is a conflict now - mind bringing it up to date?

gatorsmile commented:
sure, let me do it tonight.

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge branch 'IntersectBySemiJoinMerged' into IntersectBySemiJoinMergedNew

gatorsmile committed:
fix failed test cases

gatorsmile commented:
When resolving the conflicts, I realized the multi-children `Union` might have duplicate `exprId`. So far, I did not add a function to de-duplicate them. This is not a trivial work, if needed. When the children has hundreds of nodes, it is infeasible to use the current per-pair de-duplication. That means, we need to rewrite the whole function `dedup`. 

Let me know if we need to open a separate PR to do it now. So far, unlike `Intersect`, we did not hit any issue even if there exist duplicate `exprId` values. Thanks!

SparkQA commented:
**[Test build #49936 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49936/consoleFull)** for PR 10630 at commit [`6a7979d`](https://github.com/apache/spark/commit/6a7979d35d5cf89d9cd14eb6d6a2a4418a44a669).

SparkQA commented:
**[Test build #49936 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49936/consoleFull)** for PR 10630 at commit [`6a7979d`](https://github.com/apache/spark/commit/6a7979d35d5cf89d9cd14eb6d6a2a4418a44a669).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49936/
Test PASSed.

marmbrus commented:
I don't think its a problem for there to be conflicting attribute ids for set operations, this is because only on childs attribute references need to be propagated up (unlike with a join).

gatorsmile commented:
Yeah, agree! Thank you!

yhuai commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -125,17 +128,15 @@ object EliminateSerialization extends Rule[LogicalPlan] {
  * Pushes certain operations to both sides of a Union, Intersect or Except operator.
+=======
+ * Pushes certain operations to both sides of a Union or Except operator.
+>>>>>>> IntersectBySemiJoinMerged
```
Seems we need to remove this.

gatorsmile replied:
yeah, sure, will do.

yhuai commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1055,6 +1045,22 @@ object ReplaceDistinctWithAggregate extends Rule[LogicalPlan] {
+ *   ==>  SELECT DISTINCT a1, a2 FROM Tab1 LEFT SEMI JOIN Tab2 ON a1<=>b1 AND a2<=>b2
+ * }}}
+ */
+object ReplaceIntersectWithSemiJoin extends Rule[LogicalPlan] {
```
I think we need to add a comment at here to mention that this rewrite is just for `INTERSECT` (i.e. `INTERSECT DISTINCT`) and is not applicable to `INTERSECT ALL`.

gatorsmile replied:
Yeah, will do it. Actually, I will also implement `INTERSECT ALL` after this one. Thanks!

gatorsmile committed:
Merge remote-tracking branch 'upstream/master' into IntersectBySemiJoinMergedNew

gatorsmile committed:
address comments.

SparkQA commented:
**[Test build #50176 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50176/consoleFull)** for PR 10630 at commit [`e566d79`](https://github.com/apache/spark/commit/e566d79b12ab1d9ed9e3124fec8290cdba99549b).

SparkQA commented:
**[Test build #50176 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50176/consoleFull)** for PR 10630 at commit [`e566d79`](https://github.com/apache/spark/commit/e566d79b12ab1d9ed9e3124fec8290cdba99549b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50176/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -111,6 +113,7 @@ object SamplePushDown extends Rule[LogicalPlan] {
 }
 
 /**
+<<<<<<< HEAD
```
remove this

gatorsmile replied:
will do.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -388,57 +445,18 @@ class Analyzer(
+          i.copy(right = dedupRight(left, right))
+        case e @ Except(left, right) =>
+          e.copy(right = dedupRight(left, right))
+        case cg: CoGroup =>
```
do we have test cases to prove that duplicated expression IDs in set operations will cause problems?

gatorsmile replied:
After we convert Intersect to Join, the query can return a wrong answer due to the pushdown of Join predicate in Optimizer. That is why I did the change. Based on the comments https://github.com/apache/spark/pull/10630#discussion_r49227379  , I also tried to resolve this issue in the other operators.

cloud-fan replied:
If we replace intersect with left-semi join during analysis, then we can get de-duplication for free(the rule for join can do the work), will it help here?

gatorsmile replied:
In this case, it should work! Let me know if we should deduplicate the expression IDs for the other operators. Thanks!

cloud-fan replied:
For other operators, can we construct test cases that can make them fail without de-duplication? If we can, then we should create a JIRA and fix it in another PR.

gatorsmile replied:
So far, I am unable to create such test cases. Let me revert it back and enable de-duplication for Join only. Thanks!

gatorsmile replied:
@cloud-fan After code changes, I just realized the problem. If we replace intersect with left-semi join during analysis, it does not work. We have to de-duplicate the left and the right children at first; otherwise, the tree will be like
```
== Analyzed Logical Plan ==
id: int
Distinct
+- Join LeftSemi, Some((id#1 <=> id#1))
   :- Sample 0.0, 0.4, false, 1
   :  +- Sort [id#1 ASC], false
   :     +- Project [_1#0 AS id#1]
   :        +- LogicalRDD [_1#0], MapPartitionsRDD[2] at apply at Transformer.scala:22
   +- Sample 0.4, 1.0, false, 1
      +- Sort [id#4 ASC], false
         +- Project [_1#0 AS id#4]
            +- LogicalRDD [_1#0], MapPartitionsRDD[2] at apply at Transformer.scala:22
```
The Join condition is invalid. Should we keep the rule in the optimizer? 

BTW, I will remove the `Except` and `CoGroup`.

cloud-fan replied:
ah makes sense, let's add some comment to explain this :)

gatorsmile replied:
Will do. : )

gatorsmile committed:
address comments.

SparkQA commented:
**[Test build #50257 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50257/consoleFull)** for PR 10630 at commit [`3be78c4`](https://github.com/apache/spark/commit/3be78c4197519edb4659d8064468125b5223dd29).

SparkQA commented:
**[Test build #50257 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50257/consoleFull)** for PR 10630 at commit [`3be78c4`](https://github.com/apache/spark/commit/3be78c4197519edb4659d8064468125b5223dd29).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50257/
Test FAILed.

gatorsmile committed:
fixed the failed cases.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50298/
Test FAILed.

gatorsmile commented:
retest this please

SparkQA commented:
**[Test build #50313 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50313/consoleFull)** for PR 10630 at commit [`e51de8f`](https://github.com/apache/spark/commit/e51de8f98b25281b338ca57f0f13645b626c7673).

SparkQA commented:
**[Test build #50313 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50313/consoleFull)** for PR 10630 at commit [`e51de8f`](https://github.com/apache/spark/commit/e51de8f98b25281b338ca57f0f13645b626c7673).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `abstract class SetOperation(left: LogicalPlan, right: LogicalPlan) extends BinaryNode`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50313/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -214,12 +214,22 @@ trait CheckAnalysis {
+            val conflictingAttributes = j.left.outputSet.intersect(j.right.outputSet)
             failAnalysis(
               s"""
-                 |Failure when resolving conflicting references in Join:
```
now we can keep this message as it only checks join :)

MaxGekk replied:
Can users observe the error? or it can be considered as an internal errors? BTW, we are about to convert it to an internal error in the PR: https://github.com/apache/spark/pull/41476

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -214,12 +214,22 @@ trait CheckAnalysis {
+            failAnalysis(
+              s"""
+                 |Failure when resolving conflicting references
+                 |in operator ${operator.simpleString}:
```
same here, we could say `Intersect` directly

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala:
```diff
@@ -315,4 +315,7 @@ abstract class BinaryNode extends LogicalPlan {
   override def children: Seq[LogicalPlan] = Seq(left, right)
+
+  override lazy val resolved: Boolean =
+    expressions.forall(_.resolved) && childrenResolved
```
why override this?

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceOperatorSuite.scala:
```diff
@@ -0,0 +1,46 @@
+        ReplaceIntersectWithSemiJoin) :: Nil
+  }
+
+  test("replace Intersect with Left-semi Join") {
```
also move the test for `ReplaceDistinctWithAggregate` here?

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisErrorSuite.scala:
```diff
@@ -270,7 +270,8 @@ class AnalysisErrorSuite extends AnalysisTest {
       SimpleAnalyzer.checkAnalysis(join)
     }
-    assert(error.message.contains("Failure when resolving conflicting references in Join"))
+    assert(error.message.contains("Failure when resolving conflicting references\n" +
```
we can revert this after revert the error message change.

cloud-fan commented:
LGTM. we can merge it first and @gatorsmile can address remaining comments in a follow-up PR.

rxin commented:
This is not that big. Let's just do it together here.

gatorsmile committed:
addressed comments.

gatorsmile commented:
Thank you! Just cleaned the codes. : )

cloud-fan commented:
LGTM, pending test

SparkQA commented:
**[Test build #50368 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50368/consoleFull)** for PR 10630 at commit [`b600089`](https://github.com/apache/spark/commit/b600089d4736e9768a8968fb4dead08d28014ac1).

SparkQA commented:
**[Test build #50368 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50368/consoleFull)** for PR 10630 at commit [`b600089`](https://github.com/apache/spark/commit/b600089d4736e9768a8968fb4dead08d28014ac1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50368/
Test PASSed.

rxin commented:
Thanks - I'm going to merge this.
